### ISSUE TITLE
Get updates for unread channel messages without Flood prevention

### ISSUE THREAD

**Issue by paulbir** [sochix/TLSharp#661]
How do I get updates for new channel messages without bumping flood limits?
First I tried this:
```
            while (true)
            {
                Task.Delay(500).Wait();
                TLState state = client.SendRequestAsync<TLState>(new TLRequestGetState()).Result;
                //if (state.UnreadCount <= 0) continue;

                TLRequestGetDifference req = new TLRequestGetDifference { Date = state.Date, Pts = state.Pts, Qts = state.Qts, Sequence = state.Seq + 1 };
                TLDifference diff = client.SendRequestAsync<TLAbsDifference>(req).Result as TLDifference;

                if (diff != null)
                {
                    foreach (TLUpdateNewChannelMessage upd in diff.OtherUpdates.OfType<TLUpdateNewChannelMessage>())
                        Console.WriteLine(((TLMessage)upd.Message).Message);

                    foreach (TLChannel ch in diff.Chats.OfType<TLChannel>().Where(x => !x.Left))
                    {
                        TLInputChannel ich = new TLInputChannel { ChannelId = ch.Id, AccessHash = ch.AccessHash ?? 0 };
                        TLRequestReadHistory readed = new TLRequestReadHistory { Channel = ich, MaxId = -1 };
                        client.SendRequestAsync<bool>(readed).Wait();
                    }
                }
            }
```

But `state.UnreadCount > 0` only for unread messages from PeerUsers. And even if `state.UnreadCount > 0`, `client.SendRequestAsync<TLAbsDifference>` always returns `TLDifferenceEmpty`.

Also I tried:
```
            while (true)
            {
                Task.Delay(2000).Wait();
                var dialogs = client.GetUserDialogsAsync().Result as TLDialogs;
                if (dialogs == null) continue;

                foreach (TLDialog dialog in dialogs.Dialogs.Where(lambdaDialog => lambdaDialog.Peer is TLPeerChannel && lambdaDialog.UnreadCount > 0))
                {
                    TLPeerChannel peer = (TLPeerChannel)dialog.Peer;
                    TLChannel channel = dialogs.Chats.OfType<TLChannel>().First(lambdaChannel => lambdaChannel.Id == peer.ChannelId);
                    TLInputPeerChannel target = new TLInputPeerChannel { ChannelId = channel.Id, AccessHash = channel.AccessHash ?? 0 };

                    TLChannelMessages hist = (TLChannelMessages)client.GetHistoryAsync(target, 0, -1, dialog.UnreadCount).Result;

                    var users = hist.Users.OfType<TLUser>().ToList();
                    foreach (TLMessage message in hist.Messages.OfType<TLMessage>())
                    {
                        TLUser sentUser = users.Single(lambdaUser => lambdaUser.Id == message.FromId);
                        Console.WriteLine($"{sentUser.FirstName} {sentUser.LastName} {sentUser.Username}: {message.Message}");
                    }

                    TLInputChannel channelToMarkRead = new TLInputChannel { ChannelId = target.ChannelId, AccessHash = target.AccessHash };
                    int firstUnreadMessageId = ((TLMessage)hist.Messages.First()).Id;
                    TLRequestReadHistory markHistoryAsReadRequest = new TLRequestReadHistory
                    {
                        Channel = channelToMarkRead,
                        MaxId = -1,
                        ConfirmReceived = true,
                        Dirty = true,
                        MessageId = firstUnreadMessageId,
                        Sequence = dialog.UnreadCount
                    };

                    bool isMarked = client.SendRequestAsync<bool>(markHistoryAsReadRequest).Result;
                }
            }
```
It shows me unread messages, but has two problems.

1. It causes Flood prevention exception after several iterations if there are no new messages.
2. It causes InvalidCastException on second iteration if there are new messages. 
Here what happens:
- I call `client.SendRequestAsync<bool>(markHistoryAsReadRequest)` to mark messages as read. 
- It always returns `False`, despite messages have been marked as read.
- On the next iteration I call `var dialogs = client.GetUserDialogsAsync().Result as TLDialogs;` again.
- It causes InvalidCastException because `SendRequestAsync<TLAbsDialogs>`  is called inside `GetUserDialogsAsync`. But it receives `TLBoolTrue ` as a result. And tries to cast it to `TLAbsDialogs` which fails.

**Comment by knocte** [comment]
IMO the problem here is that you're using polling to get new messages.

I haven't investigated this yet but there should be a way for the server to send events to the client once connected.

**Comment by paulbir** [comment]
In some similar issue I saw an advice to try to use `TLRequestRegisterDevice` request, but I didn't understand how to use it. Where should I get device token and how do I get push messages? And can I even use it with windows desktop?

**Comment by paulbir** [comment]
In my second approach with `GetUserDialogsAsync` I could try to play around with random delays between requests. But as I mentioned it crashes on second iteration. Can I use some workaround to skip this `TLBoolTrue` result?

**Comment by knocte** [comment]
> Where should I get device token and how do I get push messages? And can I even use it with windows desktop?

I don't know how to answer this sorry. Maybe the best way to figure this out is look at how TelegramDesktop's official app is built, see what's their even reception approach there, and "translate" it to C#.

> But as I mentioned it crashes on second iteration.

If you're referring to InvalidCastException, those are definitely TLSharp bugs. Please contribute a PullRequest to fix the InvalidCastException.

Thanks

**Comment by paulbir** [comment]
So finally it seems to work so far. I'll post my workarounds here just in case. I'm not going to pullrequest because my changes to TLSharp source are dirty hacks and workarounds just for my case.
In `TLRequestGetDialogs` change
`Response = (Messages.TLAbsDialogs)ObjectUtils.DeserializeObject(br);`
to
`Response = ObjectUtils.DeserializeObject(br) as Messages.TLAbsDialogs;`

In `TLRequestGetHistory` change
`Response = (Messages.TLAbsMessages)ObjectUtils.DeserializeObject(br);`
to
`Response = ObjectUtils.DeserializeObject(br) as Messages.TLAbsMessages;`

So no exception will be thrown in case of casting to wrong type.

And my code:
```
            var rnd = new Random();

            while (true)
            {
                Task.Delay(rnd.Next(3000, 6000)).Wait();
                var dialogs = client.GetUserDialogsAsync().Result as TLDialogs;

                if (dialogs == null)
                {
                    Task.Delay(rnd.Next(3000, 6000)).Wait();
                    continue;
                }

                foreach (TLDialog dialog in dialogs.Dialogs.Where(lambdaDialog => lambdaDialog.Peer is TLPeerChannel && lambdaDialog.UnreadCount > 0))
                {
                    TLPeerChannel peer = (TLPeerChannel)dialog.Peer;
                    TLChannel channel = dialogs.Chats.OfType<TLChannel>().First(lambdaChannel => lambdaChannel.Id == peer.ChannelId);
                    TLInputPeerChannel target = new TLInputPeerChannel { ChannelId = channel.Id, AccessHash = channel.AccessHash ?? 0 };

                    Task.Delay(rnd.Next(3000, 6000)).Wait();
                    TLChannelMessages hist = (TLChannelMessages)client.GetHistoryAsync(target, 0, -1, dialog.UnreadCount).Result;
                    if (hist == null) continue;

                    var users = hist.Users.OfType<TLUser>().ToList();
                    var messages = hist.Messages.OfType<TLMessage>().ToList();

                    foreach (TLMessage message in messages)
                    {
                        TLUser sentUser = users.Single(lambdaUser => lambdaUser.Id == message.FromId);
                        Console.WriteLine($"{channel.Title} {sentUser.FirstName} {sentUser.LastName} {sentUser.Username}: {message.Message}");
                    }

                    TLInputChannel channelToMarkRead = new TLInputChannel { ChannelId = target.ChannelId, AccessHash = target.AccessHash };
                    var firstAbsMessage = hist.Messages[0];
                    int firstUnreadMessageId;
                    if (firstAbsMessage is TLMessage) firstUnreadMessageId = ((TLMessage) firstAbsMessage).Id;
                    else if (firstAbsMessage is TLMessageService) firstUnreadMessageId = ((TLMessageService)firstAbsMessage).Id;
                    else continue;

                    TLRequestReadHistory markHistoryAsReadRequest = new TLRequestReadHistory
                    {
                        Channel = channelToMarkRead,
                        MaxId = -1,
                        ConfirmReceived = true,
                        Dirty = true,
                        MessageId = firstUnreadMessageId,
                        Sequence = dialog.UnreadCount
                    };

                    Task.Delay(rnd.Next(3000, 6000)).Wait();
                    client.SendRequestAsync<bool>(markHistoryAsReadRequest).Wait();

                    Console.WriteLine("Mark messages as read");
                }
            }
```
You can see that I just skip execution if request result is null. Also I added random delay before each request. So far so good.

**Comment by TgWindoz** [comment]
`		private static Task<TLState> offset;

		private static async Task<TLState> GetOffset() => await Program.client.SendRequestAsync<TLState>(new TLRequestGetState());

		static void Worker()
		{

		        if (Program.client.IsUserAuthorized())
                {
				    offset = Task.Run(GetOffset);
                    while (true)
                    {
                        Thread.Sleep(500);
                        Task getUpdates = GetUpdates(offset.Result.Date, offset.Result.Pts, offset.Result.Qts);
                        getUpdates.Wait();
                    }
				}
		}

		private static async Task GetUpdates(int date, int pts, int qts)
		{
            var req = new TLRequestGetDifference() { Date = date, Pts = pts, Qts = qts};
            if (await Program.client.SendRequestAsync<TLAbsDifference>(req) is TLDifference diff)
            {
                foreach (var upd in diff.NewMessages)
                {
                    offset.Result.Pts++;
                    var msg = upd as TLMessage;
                    Console.WriteLine(msg.Message);
                }
            }
            await client.SendPingAsync();
        }`

**Comment by TgWindoz** [comment]
```
		private static Task < TLState > offset;

		private static async Task < TLState > GetOffset() => await Program.client.SendRequestAsync < TLState > (new TLRequestGetState());

		static void Worker() {

		 if (Program.client.IsUserAuthorized()) {
		  offset = Task.Run(GetOffset);
		  while (true) {
		   Thread.Sleep(500);
		   Task getUpdates = GetUpdates(offset.Result.Date, offset.Result.Pts, offset.Result.Qts);
		   getUpdates.Wait();
		  }
		 }
		}

		private static async Task GetUpdates(int date, int pts, int qts) {
		 var req = new TLRequestGetDifference() {
		  Date = date, Pts = pts, Qts = qts
		 };
		 if (await Program.client.SendRequestAsync < TLAbsDifference > (req) is TLDifference diff) {
		  foreach(var upd in diff.NewMessages) {
		   offset.Result.Pts++;
		   var msg = upd as TLMessage;
		   Console.WriteLine(msg.Message);
		  }
		 }
		 await client.SendPingAsync();
		}
```

**Comment by TgWindoz** [comment]
```
private static Task<TLState> offset;

private static async Task<TLState> GetOffset() => await Program.client.SendRequestAsync<TLState>(new TLRequestGetState());

static void Worker()
{

	if (Program.client.IsUserAuthorized())
    {
		offset = Task.Run(GetOffset);
		while (true)
        {
			Thread.Sleep(500);
			Task getUpdates = GetUpdates(offset.Result.Date, offset.Result.Pts, offset.Result.Qts);
			getUpdates.Wait();
        }
	}
}

private static async Task GetUpdates(int date, int pts, int qts)
{
    var req = new TLRequestGetDifference() { Date = date, Pts = pts, Qts = qts};
	if (await Program.client.SendRequestAsync<TLAbsDifference>(req) is TLDifference diff)
	{
		foreach (var upd in diff.NewMessages)
		{
			offset.Result.Pts++;
			var msg = upd as TLMessage;
			Console.WriteLine(msg.Message);
		}
	}
	await client.SendPingAsync();
}
```

**Comment by TgWindoz** [comment]
```
private static Task<TLState> offset;

private static async Task<TLState> GetOffset() => await Program.client.SendRequestAsync<TLState>(new TLRequestGetState());

static void Worker()
{

    if (Program.client.IsUserAuthorized())
    {
        offset = Task.Run(GetOffset);
        while (true)
        {
            Thread.Sleep(500);
            Task getUpdates = GetUpdates(offset.Result.Date, offset.Result.Pts, offset.Result.Qts);
            getUpdates.Wait();
        }
    }
}

private static async Task GetUpdates(int date, int pts, int qts)
{
    var req = new TLRequestGetDifference() { Date = date, Pts = pts, Qts = qts};
    if (await Program.client.SendRequestAsync<TLAbsDifference>(req) is TLDifference diff)
    {
        foreach (var upd in diff.NewMessages)
        {
            offset.Result.Pts++;
            var msg = upd as TLMessage;
            Console.WriteLine(msg.Message);
        }
    }
    await client.SendPingAsync();
}
```

**Comment by paulbir** [comment]
As you can see in my first post `TLRequestGetDifference` always returns `TLDifferenceEmpty`

**Comment by TgWindoz** [comment]
@paulbir 
>`TLState state = client.SendRequestAsync<TLState>(new TLRequestGetState()).Result;`

you do not have to get TLState every time, but you have to get it only once before the while loop and then increase by 1 pts

**Comment by paulbir** [comment]
My mistake with copy paste. It was my late attempt with state. I tried without setting state each time.

**Comment by TgWindoz** [comment]
send current source @paulbir

**Comment by ppanhoto78** [comment]
@paulbir have you implemented PUSH notifications?? I'm checking out the source code to do it myself. apparently, there are portions of already done:

        private bool HandleUpdate(ulong messageId, int sequence, BinaryReader messageReader)
        {
            return false;

            /*
			try
			{
				UpdatesEvent(TL.Parse<Updates>(messageReader));
				return true;
			}
			catch (Exception e)
			{
				logger.warning("update processing exception: {0}", e);
				return false;
			}
			*/
        }

**Comment by ppanhoto78** [comment]
Is it possible to receive updates without polling?

**Comment by knocte** [comment]
@ppanhoto78 it should be possible because the official TelegramDesktop client does it. You could check their source code and publish the approach here.

**Comment by paulbir** [comment]
@TgWindoz, Here is the source I tried last time:

```
            while (!token.IsCancellationRequested)
            {
                await Task.Delay(500, token);
                TLState state = await client.SendRequestAsync<TLState>(new TLRequestGetState());

                //if (state.UnreadCount <= 0) continue;

                var req = new TLRequestGetDifference {Date = state.Date, Pts = state.Pts, Qts = state.Qts};
                TLAbsDifference tmpdiff = await client.SendRequestAsync<TLAbsDifference>(req);

                if (!(tmpdiff is TLDifference)) continue;
                TLDifference diff = tmpdiff as TLDifference;

                foreach (TLUpdateNewChannelMessage upd in diff.OtherUpdates.OfType<TLUpdateNewChannelMessage>())
                    Console.WriteLine(((TLMessage)upd.Message).Message);

                foreach (TLChannel ch in diff.Chats.OfType<TLChannel>().Where(x => !x.Left))
                {
                    var inputChannel = new TLInputChannel {ChannelId = ch.Id, AccessHash = ch.AccessHash ?? 0};
                    var markHistoryAsReadRequest = new TLRequestReadHistory {Channel = inputChannel, MaxId = -1};
                    await client.SendRequestAsync<bool>(markHistoryAsReadRequest);
                }
            }
```
`tmpdiff` is used to watch the real type of TLAbsDifference return value while debugging. And it is always `TLDifferenceEmpty` even though I have unread messages at the moment.

**Comment by pisitch** [comment]
I try so many times and I found that
```TLState state = await client.SendRequestAsync<TLState>(new TLRequestGetState()); ```
This method alway return "unread = 0" for Channel but Chat and Group Chat is ok.

**Comment by ppanhoto78** [comment]
I did my version of an event handling engine. I'm not happy with it, though. It can be checked out at 
https://github.com/ppanhoto78/TLSharp/

**Comment by pisitch** [comment]
@ppanhoto78  Thank. You code guide me something.

**Comment by knocte** [comment]
@ppanhoto78 why are you not happy with it?

**Comment by ppanhoto78** [comment]
That's probably me being too picky about the solution.

1. It looks "hacky" and not look very idiomatic. 

2. that event loop it imposes on the code `for (;;) { client.WaitForEventsAsync(); }` looks uncanny -- that loop definitely should be a function to be awaited at the end of initialisation. 
  It's purpose was to have the library always ready to receive data and I did that to make it compatible with the Request/Response model used generally in the library.
  It imposes on the application that every event handler must call `SendRequestAsync` on the same "pipeline" as the main loop or we might end with a problem.

3. It uses magic numbers that may be changed if the protocol changes:

                case 0xe317af7e:
                case 0x914fbf11:
                case 0x16812688:
                case 0x78d4dec1:
                case 0x725b04c3:
                case 0x74ae4240:
                case 0x11f1331c:
                    return HandleUpdate(code, sequence, messageReader, request);

  This was already in the original code but HandleUpdate was simply commented out. It would be good if the generator added a compilation time constant for each constructor so the switch would be `case TLShortUpdateMessage.Constructor:` and get updated if the protocol changes.

4. I mixed blocking and asynchronous code (to implement event handling from the application side)
  Probably my limitation but I wasn't able to make a delegate async (don't even know if it's possible). This is an important point because any request called inside the callback (e.g. mark history read on reply to a short update message) must be blocking to avoid competing with the loop asynchronous loop (same as item 2)

**Comment by pisitch** [comment]
@ppanhoto78 Hello, I tried your version of TLsharp. I try to integrate it to my project and coded like this below to handle receiving update instead of `for (;;) { client.WaitForEventsAsync(); }`

``` private async void StartUpdateChat()
        {
            try
            {
                await Client.WaitEventAsync();
            }
            catch
            {
                await Task.Delay(1000);
                await Client.ConnectAsync(true);
            }
            finally
            {
                StartUpdateChatHistory();
            }
        }```
and in a few sec without do nothing, I got "Couldn't read the packet length" exception.
I guess this would interrupt another part of stream. I try to figure it out but it's quite difficult for me

**Comment by paulbir** [comment]
@pisitch:
I get "Couldn't read the packet length" exception sometimes after different requests with original TLsharp. It happens in my polling code I posted before.

**Comment by knocte** [comment]
> That's probably me being too picky about the solution.

Probably. Is this solution not using polling at all? I mean, does it receive a message from the server? If yes, please propose it as a PullRequest, and we'll slowly improve it in the review process until we merge it.

**Comment by pisitch** [comment]
I just found a bug in TCPTranspot -> Receieve(). The stream sometimes return byte length 0 and after, it will not return 4 length again until reestablish all TelegramClient.

**Comment by knocte** [comment]
@pisitch can you contribute a pull request to fix it?

**Comment by pisitch** [comment]
I knew the cause of "Couldn't read the packet length". It happen because when the stream has no data received, the stream view return 0 length of byte that is why it throw this exception.

I try to fix this issue it isn't be easy.

**Comment by ppanhoto78** [comment]
@knocte  That's correct: I'm not using polling to receive server updates. Don't know yet if bugs may have been introduced. I'll create the pull request but I think of that as a starting point for a robust API.

**Comment by knocte** [comment]
Sounds good! Great work.

**Comment by Marjani** [comment]
#679
